### PR TITLE
Don't capture AsyncLocals into HttpConnectionPools timer

### PR DESCRIPTION
**Pull request by benaadams**
Resolves https://github.com/dotnet/corefx/issues/26067

### CODE REVIEW

**Comment by benaadams** [comment]
OSX
```
Output: Discovering: System.ComponentModel.Composition.Tests
Output: Discovered:  System.ComponentModel.Composition.Tests
Output: Starting:    System.ComponentModel.Composition.Tests
Output: requiredService: System.ComponentModel.Composition.Hosting.CompositionContainer+CompositionServiceShim
Output: optionalService: System.ComponentModel.Composition.Hosting.CompositionContainer+CompositionServiceShim
Output: /Users/dotnet-bot/dotnetbuild/work/917a3b49-3f41-4ab3-a05e-d9642823058b/Work/440e3a6f-a136-4a2d-8ba6-710aed01df21/Unzip/RunTests.sh: line 87:
 45954 Segmentation fault: 11  (core dumped) $RUNTIME_PATH/dotnet xunit.console.netcore.exe
 System.ComponentModel.Composition.Tests.dll -xml testResults.xml -notrait Benchmark=true
 -notrait category=nonnetcoreapptests -notrait category=nonosxtests -notrait category=OuterLoop -notrait category=failing
Output: Traceback (most recent call last):
Output:   File "/Users/dotnet-bot/.dumpling/dumpling.py", line 33, in <module>
Output:     import psutil
Output: ImportError: No module named psutil
Output: Trying to find crash dumps for project: System.ComponentModel.Composition.Tests
Output: Uploading dump file: /Users/dotnet-bot/dotnetbuild/work/917a3b49-3f41-4ab3-a05e-d9642823058b/Work/440e3a6f-a136-4a2d-8ba6-710aed01df21/Unzip/core.45954
Output: ~/dotnetbuild/work/917a3b49-3f41-4ab3-a05e-d9642823058b/Work/440e3a6f-a136-4a2d-8ba6-710aed01df21/Unzip
Output: Finished running tests. End time=17:15:53. Return value was 139
Exit Code: 139
```
@dotnet-bot test OSX x64 Debug Build

**Comment by benaadams** [comment]
NETFX
```
17:19:30   xUnit.net Console Runner (64-bit Desktop .NET 4.0.30319.42000)
17:19:31     Discovering: System.IO.Compression.Tests
17:19:31     Discovered:  System.IO.Compression.Tests
17:19:31     Starting:    System.IO.Compression.Tests
17:19:34       System.IO.Compression.GzipStreamUnitTests.Dispose_WithUnfinishedWriteAsync [FAIL]
17:19:34         Assert.False() Failure
17:19:34         Expected: False
17:19:34         Actual:   True
17:19:34         Stack Trace:
17:19:34              at System.IO.Compression.CompressionStreamUnitTestBase.<Dispose_WithUnfinishedWriteAsync>d__5.MoveNext()
17:19:34           --- End of stack trace from previous location where exception was thrown ---
17:19:34              at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
17:19:34              at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
17:19:34           --- End of stack trace from previous location where exception was thrown ---
17:19:34              at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
17:19:34              at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
17:19:34           --- End of stack trace from previous location where exception was thrown ---
17:19:34              at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
17:19:34              at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
17:19:36     Finished:    System.Collections.Concurrent.Tests
```
@dotnet-bot test NETFX x86 Release Build

File: src/System.Net.Http/src/System/Net/Http/Managed/HttpConnectionPools.cs [hunk 1]
```diff
@@ -39,7 +39,12 @@ public HttpConnectionPools(int maxConnectionsPerServer)
             _maxConnectionsPerServer = maxConnectionsPerServer;
             _pools = new ConcurrentDictionary<HttpConnectionKey, HttpConnectionPool>();
             // Start out with the timer not running, since we have no pools.
+
+            // Don't capture the current ExecutionContext and its AsyncLocals onto the timer causing them to live forever
+            ExecutionContext.SuppressFlow();
```
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/Managed/HttpConnectionPools.cs]: This needs to check whether flow was already suppressed, and then only suppress and restore flow if it wasn't already.

Comment (benaadams) [reply src/System.Net.Http/src/System/Net/Http/Managed/HttpConnectionPools.cs]: Fixed

**Pull request merged**